src/assets/scripts/airport/Arrival/ArrivalBase.js

```diff
@@ -61,150 +213,205 @@ export default class ArrivalBase {
-            if (_has(this.airport.stars, pieces[i])) {
-                star = pieces[i];
-
-                if (i > 0) {
-                    entry = pieces[i - 1];
-                }
+        // find last fix along STAR that is outside of airspace, ie: next fix is within airspace
+        // distance between closest fix outside airspace and airspace border in nm
+        let extra = 0;
+        let totalDistance = 0;
+        const waypointModelList = this.airport.findWaypointModelsForStar(
+            this.activeRouteModel.base,
+            this.activeRouteModel.origin,
+            this.airport.runway
+        );
+
+        for (let i = 0; i < waypointModelList.length; i++) {
+            const waypoint = waypointModelList[i];
+            const waypointPosition = waypoint.position;
+            let previousWaypoint = waypoint;
+            let previousPosition = waypoint.position;
+
+            if (i > 0) {
+                previousWaypoint = waypointModelList[i - 1];
+                previousPosition = previousWaypoint.position;
             }
-        }
 
-        // Find the last fix that's outside the airspace boundary
-        const fixes = this.airport.getSTAR(star, entry, runway);
-        // FIXME: this is never used. is it needed?
-        // const lastFix = fixes[0][0];
-        // distance between closest fix outside a/s and a/s border, nm
-        let extra = 0;
+            if (isWithinAirspace(this.airport, waypointPosition) && i > 0) {
+                extra = nm(calculateDistanceToBoundary(this.airport, previousPosition));
 
-        for (const i in fixes) {
-            const fix = fixes[i][0];
-            const pos = this.airport.fixes[fix].position;
-            const fix_prev = (i > 0)
-                ? fixes[i - 1][0]
-                : fix;
-            const pos_prev = (i > 0)
-                ? this.airport.fixes[fix_prev].position
-                : pos;
-
-            if (inAirspace(this.airport, pos)) {
-                if (i >= 1) {
-                    extra = nm(dist_to_boundary(this.airport, pos_prev));
-                    break;
-                }
-            } else {
-                // calculate distance between fixes
-                fixes[i][2] = nm(distance2d(pos_prev, pos));
+                continue;
             }
+
+            totalDistance += waypoint.distanceFromPreviousWaypoint;
         }
 
-        // Determine spawn offsets
-        const spawn_offsets = [];
-        // distance between succ. arrivals, nm
-        const entrail_dist = this.speed / this.frequency;
-        // TODO: replace with _map
-        const dist_total = array_sum($.map(fixes, (v) => {
-            return v[2];
-        })) + extra;
-
-        for (let i = entrail_dist; i < dist_total; i += entrail_dist) {
-            spawn_offsets.push(i);
+        // FIXME: incluing this causes aircraft to spawn within airspace. something goofy is going on here.
+        // totalDistance += extra;
```

Comment:
```javascript
    /** Backfill STAR routes with arrivals closer than the spawn point
     ** Aircraft spawn at the first point defined in the route of the entry in
     ** "arrivals" in the airport json file. When that spawn point is very far
     ** from the airspace boundary, it obviously takes quite a while for them
     ** to reach the airspace. This function spawns (all at once) arrivals along
     ** the route, between the spawn point and the airspace boundary, in order to
     ** ensure the player is not kept waiting for their first arrival aircraft.
     */
    preSpawn: function() {
```

For each stream of arrivals, the process was:
 1. Count the distance `d` along the route between the spawn point and airspace boundary
 2. Determine the MIT (miles-in-trail) at which to spawn aircraft (`speed/frequency`)
 3. Count `n` number of miles along the route from the spawn point toward the airspace boundary, and spawn an aircraft.
 4. Repeat step 3 until `n > d`.

---

So really, I'm not sure quite what you're asking, but here's what I see:
 - old lines 145-146 are related to step 3 (more specifically, storing the positions at which to spawn aircraft)
 - new lines 301-302 are related to step 1, and add together:
    - the distance along the route from the spawn point to the last fix still outside of the airspace
    - the distance between the last fix still outside the airspace and the airspace boundary